rcli/typing.py

```diff
@@ -0,0 +1,665 @@
+# -*- coding: utf-8 -*-
+"""A module for handling with typing and type hints.
+
+Functions:
+    cast: Casts a value to a specific type.
+    get_signature: Retrieves the signature of a function.
+    get_type_hints: Gets all type hints for an object, including comment type
+        hints.
+
+Classes:
+    Bounded: A sliceable subclass of any class that raises a ValueError if the
+        initialization value is out of bounds.
+    Length: A sliceable subclass of any class that implements __len__ that
+        raises a ValueError if the length of the initialization value is out of
+        bounds.
+    Singleton: A metaclass to force a class to only ever be instantiated once.
+
+Instances:
+    NoneType: A type alias for type(None)
+"""
+
+from __future__ import absolute_import
+from __future__ import unicode_literals
+
+from typing import (  # type: ignore
+    _eval_type,
+    _ForwardRef,
+    _get_defaults
+)
+from typing import (  # noqa: F401 pylint: disable=unused-import
+    Any,
+    ByteString,
+    Callable,
+    cast as std_cast,
+    Dict,
+    Generator,
+    get_type_hints as std_get_type_hints,
+    List,
+    MutableSequence,
+    MutableSet,
+    Optional,
+    Sequence,
+    Tuple,
+    Type,
+    TypeVar,
+    Union
+)
+import collections
+import functools
+import inspect
+import logging
+import re
+import sys
+import tokenize
+import types
+
+import six
+
+from . import exceptions as exc
+from .decorators import memoize
+
+
+_LOGGER = logging.getLogger(__name__)
+
+_T = TypeVar('_T')
+
+_SEQUENCE_TYPES = (
+    list,
+    tuple,
+    set,
+    frozenset,
+    collections.deque,
+    collections.Counter,
+    MutableSequence,
+    MutableSet,
+    Tuple
+)
+
+NoneType = type(None)
+
+
+class Singleton(type):
+    """A metaclass to turn a class into a singleton."""
+
+    __instance__ = None  # type: type
+
+    def __call__(cls, *args, **kwargs):
+        # type: (*Any, **Any) -> type
+        """Instantiate the class only once."""
+        if not cls.__instance__:
+            cls.__instance__ = super(Singleton, cls).__call__(*args, **kwargs)
+        return cls.__instance__
+
+
+class _Uninstantiable(type):
```

Comment:
It's not clear to me how making this logic it's own parent class is helpful.

It _could_ be moved into `_BoundedMeta` without loss of generality.

It's possible this would make some future use-case harder to implement.
It's also arguable that this parent class aids readability somehow.